accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,258 @@
+abstractions for register allocation:
+
+```rust
+pub struct Compiler {
+  //...
+  allocator: RegisterAllocator,
+  value_stack: ValueStack,
+  //...
+}
+```
+
+The value stack is expected to keep track of the location of its values.
+A particular value can be tagged as either a:
+
+* Local: representing a function local slot (index and type). The address of the
+  local will be resolved lazily to reduce register pressure. 
+* Register
+* Constant: representing an immediate value. 
+* Memory Offset: the location of the value at a given memory offset
+
+Registers will be requested to the register allocator every time an operation
+requires it. If no registers are available, the baseline compiler will move
+locals and registers to memory, changing their tag to a memory offset,
+performing what's known as spilling, effectively freeing up registers. Spilling
+will also be performed at control flow points. To reduce the number of spills,
+the baseline compiler will also perform limited constant rematerialization.
+
+Assuming that we have an immediate at the top of the stack, emitting an add
+instruction with an immediate operand would look something like this:
+
+```rust
+let mut masm = cranelift_asm::x64::MacroAssembler::new();
+// request a general purpose register;
+// spill if none available
+let imm = self.value_stack.pop(); 
+none available let rd = self.gpr();
+masm.add(rd, imm);
+```
+
+### Integration with Wasmtime
+
+We plan to integrate the baseline compiler incrementally into Wasmtime, as an
+in-tree crate, `winch`. It will be introduced as a compile-time feature, off by
+default. Taking as a guideline [Wasmtime's tiers of
+support](https://github.com/bytecodealliance/wasmtime/pull/4479), this means
+that the baseline compiler will be introduced as a Tier 3 feature.
+
+In general, the development of the baseline compiler will be done in phases,
+each phase covering a specific set of features:
+
+| Phase | Feature                  | Feature Type        |
+|-------|--------------------------|---------------------|
+| 1     | cranelift_asm crate      | Refactoring         |
+| 1     | x64 support              | Target architecture |
+| 1     | wasi_snapshot_preview1   | WASI proposal       |
+| 1     | wasi_unstable            | WASI proposal       |
+| 1     | Multi-Memory             | Wasm proposal       |
+| 1     | Reference Types          | Wasm proposal       |
+| 1     | Epoch-based interruption | Wasmtime feature    |
+| 1     | Parallel compilation     | Wasmtime feature    |
+| 1     | Fuzzing integration      | Test coverage       |
+| 2     | Fuel                     | Wasmtime feature    |
+| 2     | SIMD                     | Wasm proposal       |
+| 2     | Memory 64                | Wasm proposal       |
+| 2     | ARM support              | Target architecture |
+| 3     | s390x                    | Target architecture |
+| 3     | Debugging integration    | Debugging           |
+
+#### Configuring compilation
```

Comment:
Probably best left for the future, but it would be nice to support tiered compilation to compile with winch by default and with cranelift when a function turns out to be hot.

Reply:
Yes, tiering and its heuristics are intentionally out of scope for this RFC.